packages/headless/src/app/search-engine/search-engine.ts

```diff
@@ -0,0 +1,129 @@
+import {SearchAPIClient} from '../../api/search/search-api-client';
+import {
+  NoopPostprocessFacetSearchResponseMiddleware,
+  NoopPostprocessQuerySuggestResponseMiddleware,
+  NoopPostprocessSearchResponseMiddleware,
+} from '../../api/search/search-api-client-middleware';
+import {buildEngine, EngineOptions, ExternalEngineOptions} from '../engine';
+import {Engine} from '../headless-engine';
+import {buildLogger} from '../logger';
+import {buildThunkExtraArguments} from '../thunk-extra-arguments';
+import {Logger} from 'pino';
+import {NoopPreprocessRequest} from '../../api/preprocess-request';
+import {NoopPreprocessRequestMiddleware} from '../../api/platform-client';
+import {debug, pipeline, search, searchHub} from '../reducers';
+import {StateFromReducersMapObject} from 'redux';
+import {updateSearchConfiguration} from '../../features/configuration/configuration-actions';
+import {
+  SearchEngineConfiguration,
+  searchEngineConfigurationSchema,
+} from './search-engine-configuration-options';
+import {executeSearch} from '../../features/search/search-actions';
+import {logInterfaceLoad} from '../../features/analytics/analytics-actions';
+import {firstSearchExecutedSelector} from '../../features/search/search-selectors';
+import {SearchAppState} from '../../state/search-app-state';
+
+const searchEngineReducers = {debug, pipeline, searchHub, search};
+type SearchEngineReducers = typeof searchEngineReducers;
+type SearchEngineState = StateFromReducersMapObject<SearchEngineReducers> &
+  Partial<SearchAppState>;
+
+export interface SearchEngine extends Engine<SearchEngineState> {
+  executeFirstSearch(): void;
+}
+
+export interface SearchEngineOptions
+  extends ExternalEngineOptions<SearchEngineState> {
+  /**
+   * The search engine configuration options.
+   */
+  configuration: SearchEngineConfiguration;
+}
+
+export function buildSearchEngine(options: SearchEngineOptions): SearchEngine {
+  const logger = buildLogger(options.loggerOptions);
+  validateConfiguration(options.configuration, logger);
+
+  const ref = {
+    renewAccessToken: () => Promise.resolve(''),
+  };
+
+  const searchAPIClient = createSearchAPIClient(
+    options.configuration,
+    logger,
+    ref
+  );
+
+  const thunkArguments = {
+    ...buildThunkExtraArguments(options.configuration, logger),
+    searchAPIClient,
+  };
+
+  const augmentedOptions: EngineOptions<SearchEngineReducers> = {
+    ...options,
+    reducers: searchEngineReducers,
+  };
+
+  const engine = buildEngine(augmentedOptions, thunkArguments);
+  ref.renewAccessToken = engine.renewAccessToken;
```

Comment:
This is the circular issue. It is more apparent inside a function, but the same thing is happening inside headless engine class.

Reply:
I would add a comment that explains this hack/why it's there/and a potential TODO to refactor/improve "eventually" 

> Fast forward 4 years and the hack is still there

Reply:
Indeed, I wanted to avoid the work-around as soon as possible so that contributors creating engines for other use-cases would not need to repeat it. I created a PR here: https://github.com/coveo/ui-kit/pull/845#partial-pull-merging